### PR TITLE
allow gateway plugins to add authorization policies

### PR DESCRIPTION
**Pull request by quoideneuf**
Draft changes to the GatewayHandler to enable addition of authorization policies from gateway plugins.

### CODE REVIEW

File: pages/gateway/GatewayHandler.inc.php [hunk 1]
```diff
@@ -10,12 +10,36 @@
...
+	 * Constructor
+	 *
+	 * @param $request PKPRequest
+	 */
+	function __construct($request) {
+		parent::__construct();
+
+		$op = $request->getRouter()->getRequestedOp($request);
+		if ($op == 'plugin') {
+			$pluginName = array_slice($request->getRouter()->getRequestedArgs($request), 0, 1)[0];
```
Comment (asmecher) [inline pages/gateway/GatewayHandler.inc.php]: Isn't this just equivalent to `array_shift($request->getRouter()->getRequestedArgs($request)`?

File: pages/gateway/GatewayHandler.inc.php [hunk 2]
```diff
@@ -10,12 +10,36 @@
...
+			$plugins = PluginRegistry::loadCategory('gateways');
+			$plugin = $plugins[$pluginName];
+
+			if (method_exists($plugin, "addPolicies")) {
```
[the remark this answers is not in the document]
Comment (asmecher) [reply pages/gateway/GatewayHandler.inc.php]: (Or perhaps a new method in the `GatewayPlugin` parent class that returns an `array()` by default.)

File: pages/gateway/GatewayHandler.inc.php (hunk 1 above)
Comment (quoideneuf) [reply pages/gateway/GatewayHandler.inc.php]: Yes.

File: pages/gateway/GatewayHandler.inc.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (quoideneuf) [reply pages/gateway/GatewayHandler.inc.php]: Ok, I will make that change, but GatewayPlugin is in lib/pkp. I will make a PR there and link them unless there's a better way. I am also wondering, should I be posting these to the master branch(es) as well?

[the remark this answers is not in the document]
Comment (asmecher) [reply pages/gateway/GatewayHandler.inc.php]: Great, please do open a PR in github.com/pkp/pkp-lib. Don't worry about opening PRs to `master` as well -- when it comes time to merge I'll take care of cherry-picking forward.

[the remark this answers is not in the document]
Comment (quoideneuf) [reply pages/gateway/GatewayHandler.inc.php]: Done - See: https://github.com/pkp/pkp-lib/pull/4677

File: pages/gateway/GatewayHandler.inc.php [hunk 3]
```diff
@@ -10,12 +10,34 @@
...
+	 * @param $request PKPRequest
+	 */
+	function __construct($request) {
+		parent::__construct();
+		$op = $request->getRouter()->getRequestedOp($request);
+		if ($op == 'plugin') {
+			$args = $request->getRouter()->getRequestedArgs($request);
+			$pluginName = array_shift($args);
+			$plugins = PluginRegistry::loadCategory('gateways');
+			$this->plugin = $plugins[$pluginName];
```
[the remark this answers is not in the document]
Comment (quoideneuf) [reply pages/gateway/GatewayHandler.inc.php]: Revised - with the corrected check for the 404

**Comment by asmecher** [comment]
Thanks, @quoideneuf! Merged, forward-ported to `master`, and ported to OMP (`master` and `stable-3_1_2`).

**Pull request merged**